# Assignment 3

In [2]:
from array import array
import numpy as np
import math

## 1a)

In [24]:
theta_to_distance = {}
thetas = np.linspace(0, 90, 500)

# Initial Conditions
speed = 55.8
y1 = 0.0


# Other constants - golf ball - D = 0.0427m, m = 0.04593kg
Cd = 0.20 # drag coefficient
area = 0.001432 # cross sectional area of projectile
grav = 9.81 # gravitional acceleration
mass = 0.04593 # mass in kg
rho = 1.225 # density of air (kg/m^3)
air_const = -0.5*Cd*rho*area/mass
Pi = math.pi

for theta in thetas:
    r1 = array('d')
    v1 = array('d')
    r = array('d')
    v = array('d')
    accel = array('d')
    
    r1.append(0)
    r1.append(y1)
    v1.append(speed*math.cos(theta*Pi/180.0))
    v1.append(speed*math.sin(theta*Pi/180.0))
    
    r.append(r1[0])
    r.append(r1[1])
    v.append(v1[0])
    v.append(v1[1])
    accel.append(0)
    accel.append(0)
    
    # Setup arrays for Euler and Euler-Cromer Methods
    re = array('d')
    ve = array('d')
    accele = array('d')
    
    re.append(r1[0])
    re.append(r1[1])
    ve.append(v1[0])
    ve.append(v1[1])
    accele.append(0)
    accele.append(0)
    
    rec = array('d')
    vec = array('d')
    accelec = array('d')
    
    rec.append(r1[0])
    rec.append(r1[1])
    vec.append(v1[0])
    vec.append(v1[1])
    accelec.append(0)
    accelec.append(0)
    
    tau = 0.02 # timestep in seconds
    maxStep = 1000000 # maximum number of steps
    
    xplot = array('d')
    yplot = array('d')
    xplote = array('d')
    yplote = array('d')
    xplotec = array('d')
    yplotec = array('d')
    xNoAir = array('d')
    yNoAir = array('d')
    time = array('d')
    timee = array('d')
    timeec = array('d')
    timeNoAir = array('d')
    for iStep in range(1,maxStep+1):
        
        t = (iStep-1)*tau
            
        xplot.append(r[0])
        yplot.append(r[1])
        xplote.append(re[0])
        yplote.append(re[1])
        xplotec.append(rec[0])
        yplotec.append(rec[1])
        xNoAir.append(r1[0]+v1[0]*t)
        yNoAir.append(r1[1]+v1[1]*t-0.5*grav*t*t)
        
        timeNoAir.append(t)
        time.append(t)
        timee.append(t)
        timeec.append(t)
        
        # Euler Method
        normVe = math.sqrt(ve[0]*ve[0]+ve[1]*ve[1])
        accele[0]=air_const*normVe*ve[0]
        accele[1]=air_const*normVe*ve[1]-grav
        re[0] = re[0] + tau*ve[0]
        re[1] = re[1] + tau*ve[1]
        ve[0] = ve[0] + tau*accele[0]
        ve[1] = ve[1] + tau*accele[1]
        
        # Euler-Cromer Method
        normVec = math.sqrt(vec[0]*vec[0]+vec[1]*vec[1])
        accelec[0]=air_const*normVec*vec[0]
        accelec[1]=air_const*normVec*vec[1]-grav
        vec[0] = vec[0] + tau*accelec[0]
        vec[1] = vec[1] + tau*accelec[1]    
        rec[0] = rec[0] + tau*vec[0]
        rec[1] = rec[1] + tau*vec[1]
        
        # MidPoint Method
        normV = math.sqrt(v[0]*v[0]+v[1]*v[1])
        accel[0]=air_const*normV*v[0]
        accel[1]=air_const*normV*v[1]-grav
        vxnew = v[0] + tau*accel[0]
        vynew = v[1] + tau*accel[1]
        r[0] = r[0] + tau*(v[0] + vxnew)/2.0
        r[1] = r[1] + tau*(v[1] + vynew)/2.0
        v[0] = vxnew
        v[1] = vynew
        if r[1] < 0 and v[1] < 0:
            xplot.append(r[0])
            yplot.append(r[1])
            xplote.append(re[0])
            yplote.append(re[1])
            xplotec.append(rec[0])
            yplotec.append(rec[1])
            time.append(tau*iStep)
            timee.append(tau*iStep)
            timeec.append(tau*iStep)
            endtime = tau*iStep
            nelements = iStep
            break
    theta_to_distance[theta] = r[0]

In [25]:
print(f"Opimal theta is ~ {np.round(max(theta_to_distance, key=theta_to_distance.get), 2)} degrees with value of {np.round(theta_to_distance[max(theta_to_distance, key=theta_to_distance.get)], 2)} meters.")

Opimal theta is ~ 40.04 degrees with value of 174.87 meters.


## 1b)

In [29]:
theta_to_distance = {}
thetas = np.linspace(0, 90, 500)

# Initial Conditions
speed = 44.704 # m/s = 100 mph
y1 = 1 # m


# Other constants - golf ball - D = 0.074m, m = 0.1455kg
Cd = 0.35 # drag coefficient from here: https://www.mdpi.com/2076-3417/12/3/1416 (ignore the first chart, its for knucke balls)
area = 0.004301 # cross sectional area of projectile
grav = 9.81 # gravitional acceleration
mass = 0.1455 # mass in kg
rho = 1.225 # density of air (kg/m^3)
air_const = -0.5*Cd*rho*area/mass
Pi = math.pi

for theta in thetas:
    r1 = array('d')
    v1 = array('d')
    r = array('d')
    v = array('d')
    accel = array('d')
    
    r1.append(0)
    r1.append(y1)
    v1.append(speed*math.cos(theta*Pi/180.0))
    v1.append(speed*math.sin(theta*Pi/180.0))
    
    r.append(r1[0])
    r.append(r1[1])
    v.append(v1[0])
    v.append(v1[1])
    accel.append(0)
    accel.append(0)
    
    # Setup arrays for Euler and Euler-Cromer Methods
    re = array('d')
    ve = array('d')
    accele = array('d')
    
    re.append(r1[0])
    re.append(r1[1])
    ve.append(v1[0])
    ve.append(v1[1])
    accele.append(0)
    accele.append(0)
    
    rec = array('d')
    vec = array('d')
    accelec = array('d')
    
    rec.append(r1[0])
    rec.append(r1[1])
    vec.append(v1[0])
    vec.append(v1[1])
    accelec.append(0)
    accelec.append(0)
    
    tau = 0.02 # timestep in seconds
    maxStep = 1000000 # maximum number of steps
    
    xplot = array('d')
    yplot = array('d')
    xplote = array('d')
    yplote = array('d')
    xplotec = array('d')
    yplotec = array('d')
    xNoAir = array('d')
    yNoAir = array('d')
    time = array('d')
    timee = array('d')
    timeec = array('d')
    timeNoAir = array('d')
    for iStep in range(1,maxStep+1):
        
        t = (iStep-1)*tau
            
        xplot.append(r[0])
        yplot.append(r[1])
        xplote.append(re[0])
        yplote.append(re[1])
        xplotec.append(rec[0])
        yplotec.append(rec[1])
        xNoAir.append(r1[0]+v1[0]*t)
        yNoAir.append(r1[1]+v1[1]*t-0.5*grav*t*t)
        
        timeNoAir.append(t)
        time.append(t)
        timee.append(t)
        timeec.append(t)
        
        # Euler Method
        normVe = math.sqrt(ve[0]*ve[0]+ve[1]*ve[1])
        accele[0]=air_const*normVe*ve[0]
        accele[1]=air_const*normVe*ve[1]-grav
        re[0] = re[0] + tau*ve[0]
        re[1] = re[1] + tau*ve[1]
        ve[0] = ve[0] + tau*accele[0]
        ve[1] = ve[1] + tau*accele[1]
        
        # Euler-Cromer Method
        normVec = math.sqrt(vec[0]*vec[0]+vec[1]*vec[1])
        accelec[0]=air_const*normVec*vec[0]
        accelec[1]=air_const*normVec*vec[1]-grav
        vec[0] = vec[0] + tau*accelec[0]
        vec[1] = vec[1] + tau*accelec[1]    
        rec[0] = rec[0] + tau*vec[0]
        rec[1] = rec[1] + tau*vec[1]
        
        # MidPoint Method
        normV = math.sqrt(v[0]*v[0]+v[1]*v[1])
        accel[0]=air_const*normV*v[0]
        accel[1]=air_const*normV*v[1]-grav
        vxnew = v[0] + tau*accel[0]
        vynew = v[1] + tau*accel[1]
        r[0] = r[0] + tau*(v[0] + vxnew)/2.0
        r[1] = r[1] + tau*(v[1] + vynew)/2.0
        v[0] = vxnew
        v[1] = vynew
        if r[1] < 0 and v[1] < 0:
            xplot.append(r[0])
            yplot.append(r[1])
            xplote.append(re[0])
            yplote.append(re[1])
            xplotec.append(rec[0])
            yplotec.append(rec[1])
            time.append(tau*iStep)
            timee.append(tau*iStep)
            timeec.append(tau*iStep)
            endtime = tau*iStep
            nelements = iStep
            break
    theta_to_distance[theta] = r[0]

In [30]:
print(f"Opimal theta is ~ {np.round(max(theta_to_distance, key=theta_to_distance.get), 2)} degrees with distance of {np.round(theta_to_distance[max(theta_to_distance, key=theta_to_distance.get)], 2)} meters.")

Opimal theta is ~ 40.4 degrees with distance of 110.13 meters.


In [33]:
print(f"Percent error between angles is {np.round((40.4 - 28)/28*100, 2)}%")

Percent error between angles is 44.29%


In [35]:
print(f"Percent error between distances is {np.round((340 - 361.3)/340*100, 2)}%")

Percent error between distances is -6.26%


Compared to the table from https://baseballsavant.mlb.com/statcast_hit_probability?year=2023&type=la, The calculated angle of 41.66 degrees is off by ~12.4 degrees from the experimental maximum of 28 degrees. The calculated max distance is roughly 361.3 feet, which is fairly close to the experimental maximum of 340 feet, off by ~6.3% (using the experimental as the expected/truth). Its interesting that the calculated max distance is relatively close to th experimental when considering that the angle had a percent error of 44.3%. This could be due to a multitude of reasons, with my primary heuristic being the physiological constraints humans might have in throwing baseballs. There is also uncertainty in the drag coefficients, as the experimental data (and model) doesn't account for spin.